### PR TITLE
FIx for DP + multicore with IPC4

### PR DESCRIPTION
**Pull request by ranj063**
replacement for #4692

### CODE REVIEW

**Comment by marc-hb** [comment]
SOFCI TEST

File: sound/soc/sof/ipc4-topology.c [hunk 1]
```diff
@@ -692,6 +693,7 @@ static int sof_ipc4_widget_setup_comp_pipeline(struct snd_sof_widget *swidget)
 	}
 
 	swidget->core = pipeline->core_id;
+	spipe->core_mask |= BIT(pipeline->core_id);
```
Comment (bardliao) [inline sound/soc/sof/ipc4-topology.c:696]: Do we need to clear core_mask?

Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:696]: This is set during topology parsing and will remain unchanged until the topology is unloaded. I don't think we need to clear the core mask

**Review by bardliao — DISMISSED** [review]

File: sound/soc/sof/ipc4-topology.c [hunk 2]
```diff
@@ -943,6 +946,10 @@ static int sof_ipc4_widget_setup_comp_process(struct snd_sof_widget *swidget)
 
 	sof_ipc4_widget_update_kcontrol_module_id(swidget);
 
+	/* set pipeline core mask to keep track of the core the DP module is scheduled to run on */
+	if ((fw_module->man4_module_entry.type & SOF_IPC4_MODULE_DP))
+		spipe->core_mask |= BIT(swidget->core);
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: Why the distinction of DP modules?
Non DP modules can only be executed on core 0?

File: sound/soc/sof/sof-audio.h [hunk 3]
```diff
@@ -535,7 +537,7 @@ int snd_sof_switch_get(struct snd_kcontrol *kcontrol,
 int snd_sof_switch_put(struct snd_kcontrol *kcontrol,
 		       struct snd_ctl_elem_value *ucontrol);
 int snd_sof_enum_get(struct snd_kcontrol *kcontrol,
-		     struct snd_ctl_elem_value *ucontrol);
+		  struct snd_ctl_elem_value *ucontrol);
```
Comment (ujfalusi) [inline sound/soc/sof/sof-audio.h]: Unrelated change.

File: sound/soc/sof/sof-audio.c [hunk 4]
```diff
@@ -85,16 +86,23 @@ static int sof_widget_free_unlocked(struct snd_sof_dev *sdev,
...
-	if (ret < 0) {
-		dev_err(sdev->dev, "error: failed to disable target core: %d for widget %s\n",
-			swidget->core, swidget->widget->name);
-		if (!err)
-			err = ret;
+	/* disable cores associated with all modules in the pipeline and clear the complete flag */
+	if (swidget->id == snd_soc_dapm_scheduler) {
+		int i;
+
+		for (i = 0; i < sdev->num_cores; i++) {
```
Comment (ujfalusi) [inline sound/soc/sof/sof-audio.c]: `for_each_set_bit(i, &spipe->core_mask, sdev->num_cores)` ?

File: sound/soc/sof/ipc4-topology.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c]: LL modules cannot be on a core different from the pipeline core set in topology

**Review by yongzhi1 — APPROVED** [review]

**Comment by ranj063** [comment]
SOFCI TEST

File: sound/soc/sof/ipc4-topology.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-topology.c:696]: what happens if the machine driver is rmmod'ed, that would unload the topology and then what happens on modprode?

File: sound/soc/sof/ipc4-topology.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-topology.c]: @ranj063 Is this correct? IIRC that's not allowed *today* but will be in the near future. It's my understanding that all restrictions on LL scheduling will be removed.

File: sound/soc/sof/sof-audio.c [hunk 5]
```diff
@@ -45,6 +45,7 @@ static int sof_widget_free_unlocked(struct snd_sof_dev *sdev,
 				    struct snd_sof_widget *swidget)
 {
 	const struct sof_ipc_tplg_ops *tplg_ops = sof_ipc_get_ops(sdev, tplg);
+	struct snd_sof_pipeline *spipe = swidget->spipe;
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:48]: typo/grammar in commit message: because the secondary core one of its widgets

File: sound/soc/sof/sof-audio.c [hunk 6]
```diff
@@ -85,16 +86,20 @@ static int sof_widget_free_unlocked(struct snd_sof_dev *sdev,
...
-	 * disable widget core. continue to route setup status and complete flag
-	 * even if this fails and return the appropriate error
-	 */
-	ret = snd_sof_dsp_core_put(sdev, swidget->core);
-	if (ret < 0) {
-		dev_err(sdev->dev, "error: failed to disable target core: %d for widget %s\n",
-			swidget->core, swidget->widget->name);
-		if (!err)
-			err = ret;
+	/* disable cores associated with all modules in the pipeline and clear the complete flag */
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c]: it's not really disable, it's update the refcount. The disable only happens if the refcount is zero.

File: sound/soc/sof/sof-audio.c [hunk 7]
```diff
@@ -85,16 +86,20 @@ static int sof_widget_free_unlocked(struct snd_sof_dev *sdev,
...
+	if (swidget->id == snd_soc_dapm_scheduler) {
+		int i;
+
+		for_each_set_bit(i, &spipe->core_mask, sdev->num_cores) {
+			ret = snd_sof_dsp_core_put(sdev, i);
+			if (ret < 0) {
+				dev_err(sdev->dev, "error: failed to disable target core: %d for pipeline %s\n",
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c]: remove error: prefix

File: sound/soc/sof/sof-audio.c [hunk 8]
```diff
@@ -165,19 +168,23 @@ static int sof_widget_setup_unlocked(struct snd_sof_dev *sdev,
 			goto use_count_dec;
 	}
 
-	/* enable widget core */
-	ret = snd_sof_dsp_core_get(sdev, swidget->core);
-	if (ret < 0) {
-		dev_err(sdev->dev, "error: failed to enable target core for widget %s\n",
-			swidget->widget->name);
-		goto pipe_widget_free;
+	/* enabled cores associated with all modules in a pipeline */
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c]: update refcount, and enable core if refcount moved from zero to one.

File: sound/soc/sof/sof-audio.c [hunk 9]
```diff
@@ -85,16 +86,20 @@ static int sof_widget_free_unlocked(struct snd_sof_dev *sdev,
...
-		dev_err(sdev->dev, "error: failed to disable target core: %d for widget %s\n",
-			swidget->core, swidget->widget->name);
-		if (!err)
-			err = ret;
+	/* disable cores associated with all modules in the pipeline and clear the complete flag */
+	if (swidget->id == snd_soc_dapm_scheduler) {
+		int i;
+
+		for_each_set_bit(i, &spipe->core_mask, sdev->num_cores) {
+			ret = snd_sof_dsp_core_put(sdev, i);
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:97]: the only part that scares me is the locking model. How is the refcount protected between get and put?

**Review by plbossart — COMMENTED** [review]
The logic sounds mostly good @ranj063, the only concerns I have are locking and a couple of comment on comments.

File: sound/soc/sof/sof-audio.c (hunk 9 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:97]: @plbossart widget setup/free is protected by the widget setup_mutex and also the dpcm mutex during hw_params/free by the core. We've never had a mutex protecting core ref counts before. Im not sure it is really needed.

File: sound/soc/sof/ipc4-topology.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c]: @plbossart I think that's correct. Let me confirm it with the firmware team

File: sound/soc/sof/sof-audio.c (hunk 9 above)
Comment (plbossart) [reply sound/soc/sof/sof-audio.c:97]: well the problem is that the dpcm mutex only applies between hw_params and hw_free *of the same PCM device*. But here there is no real correlation when you have shared BE pipelines inside the firmware. With the deferred count I think you have a larger attack surface to worry about, no?

Comment (ranj063) [reply sound/soc/sof/sof-audio.c:97]: with shared BE pipelines, we have the widget_mutex protecting the widget setup/free no?

File: sound/soc/sof/ipc4-topology.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:696]: it will get reset anyway because all IPC structures are reallocated

File: sound/soc/sof/sof-audio.c [hunk 10]
```diff
@@ -207,15 +217,22 @@ static int sof_widget_setup_unlocked(struct snd_sof_dev *sdev,
...
-		snd_sof_dsp_core_put(sdev, swidget->core);
 pipe_widget_free:
-	if (swidget->id != snd_soc_dapm_scheduler)
+	if (swidget->id != snd_soc_dapm_scheduler) {
 		sof_widget_free_unlocked(sdev, swidget->spipe->pipe_widget);
+	} else {
+		int j;
+
+		/* decrement ref count for all cores that were updated previously */
+		for_each_set_bit(j, &spipe->core_mask, sdev->num_cores) {
```
Comment (ujfalusi) [inline sound/soc/sof/sof-audio.c:230]: not sure if it would work, but in theory
```c
for_each_set_bit(j, &spipe->core_mask, i + 1) {
```
might work as well?

**Review by ujfalusi — DISMISSED** [review]

File: sound/soc/sof/ipc3-topology.c [hunk 11]
```diff
@@ -505,6 +505,7 @@ static int sof_ipc3_widget_setup_comp_mixer(struct snd_sof_widget *swidget)
 static int sof_ipc3_widget_setup_comp_pipeline(struct snd_sof_widget *swidget)
 {
 	struct snd_soc_component *scomp = swidget->scomp;
+	struct snd_sof_pipeline *spipe = swidget->spipe;
```
Comment (plbossart) [inline sound/soc/sof/ipc3-topology.c:508]: typos in commit message

they can be scheduled
Add a new field in struct

File: sound/soc/sof/ipc4-topology.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-topology.c]: but what would happen if you unconditionally set this mask - regargly of the type of module? This wouldn't break today and would be future-proof if LL modules have extra permissions

Conversely if LL modules CANNOT be on a different pipeline then we need some sort of error check.

**Review by plbossart — CHANGES_REQUESTED** [review]

File: sound/soc/sof/ipc4-topology.c (hunk 2 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.c]: @plbossart, I was also contemplating on this. It will not hurt at all to set the same bit several times and it will for sure future proof the code and it is at the same time makes it even cleaner: we are marking the core(s) that are used by the modules on the pipeline.

Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c]: @plbossart ok i've removed the restriction to limit this to only DP modules now.

File: sound/soc/sof/ipc4-topology.c [hunk 12]
```diff
@@ -943,6 +949,9 @@ static int sof_ipc4_widget_setup_comp_process(struct snd_sof_widget *swidget)
 
 	sof_ipc4_widget_update_kcontrol_module_id(swidget);
 
+	/* set pipeline core mask to keep track of the core the module is scheduled to run on */
+	spipe->core_mask |= BIT(swidget->core);
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c:953]: question: do we check if the cores used by topology are compatible with the hardware definitions?

Just being paranoid, e.g. if we use a topology defined for 4 cores on a SKU that only has 3.

**Review by plbossart — APPROVED** [review]
LGTM @ranj063, just one question below

File: sound/soc/sof/ipc4-topology.c (hunk 12 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:953]: not during topology parsing but we do check before trying to do a core_get op

**Review by yongzhi1 — APPROVED** [review]

**Review by ujfalusi — APPROVED** [review]
